WordPress/gutenberg
# Table of Contents: Try storing in post-meta

## What?
Part of #42229.
Resolves #41031.
Resolves #43595.

PR changes the storing mechanism for ToC block headings from attribute to post meta. This opens up the possibility of rendering the ToC block outside of post content (see #41173) and fixes a block's "undo quirk" issue.

The dynamic rendering allows us to skip storing permalinks in DB, which resolves #43595 and should make migrating content with the ToC block easier. You don't need to replace domains in serialized content.

> [!NOTE]  
> I've intentionally omitted post-meta revision handling from this PR. I hope that [WP#4859](https://github.com/WordPress/wordpress-develop/pull/4859) will also land in WP 6.4.

## Why?

* **Undo quirk** - The table of contents uses a derived state from the block editor and shouldn't create an undo step. Marking attribute changes as non-persistent conflicted with attribute edit coalescing and breaking the user experience.
* **Permalinks** - The page permalinks are only generated after publishing; storing them resulted in stale data.

## How?

* Now that the block uses post meta for storing the derived data, we can easily skip creating the undo step by using the `undoIgnore` option flag.
* Permalinks for heading entries are dynamic.

## Testing Instructions
1. Open a page.
2. Add a Table of Contents blocks.
3. Add headings of various levels.
4. Publish the page.
5. Confirm that ToC links work when viewing the page (See #43595).
6. Reload the page in the editor.
7. Make changes to a Heading block.
8. Confirm edit coalescing works again for Heading blocks. It doesn't undo changes letter by letter.

## Screenshots or screencast 

**Before**

https://github.com/WordPress/gutenberg/assets/240569/d3a64775-e55b-4338-9fe7-1bebd7c5edbb

**After**

https://github.com/WordPress/gutenberg/assets/240569/d54046c8-9c19-406a-b783-b2b837b1770f
